contracts/03-tamagotchi-nft/src/lib.rs

```diff
@@ -1,19 +1,178 @@
 #![no_std]
 
 #[allow(unused_imports)]
-use gstd::prelude::*;
+use gstd::{
+    prelude::*,
+    msg,
+    exec
+};
+use tamagotchi_nft_io::*;
+
+static mut TAMAGOTCHI: Option<Tamagotchi> = None;
 
 #[no_mangle]
 extern fn init() {
     // TODO: 0️⃣ Copy the `init` function from the previous lesson and push changes to the master branch
+    let tamagotchi_name: String = msg::load().expect("Error in init message");
+    let block_height = blocks_height();
+    let new_tamagotchi: Tamagotchi = Tamagotchi {
+        name: tamagotchi_name,
+        date_of_birth: block_height,
+        owner: msg::source(),
+        fed: 5000,
+        fed_block: block_height,
+        entertained: 5000,
+        entertained_block: block_height,
+        rested: 5000,
+        rested_block: block_height,
+        approved_account: None
+    };
+    unsafe {
+        TAMAGOTCHI = Some(new_tamagotchi);
+    };
+    msg::reply("successful initialization!", 0)
+        .expect("error in reply");
 }
 
 #[no_mangle]
 extern fn handle() {
     // TODO: 0️⃣ Copy the `handle` function from the previous lesson and push changes to the master branch
+    let type_message: TmgAction = msg::load()
+        .expect("error in load message");
+    // In each call all tamagotchi fields are updated
+    update_tamagotchi_fields();
+    let tamagotchi = state_mut();
+
+    match type_message {
+        TmgAction::Name => {
+            let tamagotchi_name = TmgEvent::Name(String::from(&tamagotchi.name));
+            msg::reply(tamagotchi_name, 0)
+                .expect("Error in sending tamagotchi name");
+        },
+        TmgAction::Age => {
+            let tamagotchi_age = TmgEvent::Age(tamagotchi.date_of_birth);
+            msg::reply(tamagotchi_age, 0)
+                .expect("Errorin sending tamagotchi age");
+        },
+        TmgAction::Feed => {
+            update_field(&mut tamagotchi.fed, FILL_PER_FEED);
+            msg::reply(TmgEvent::Fed, 0)
+                .expect("Error sending tamagotchi variant 'Fed'");
+        },
+        TmgAction::Play => {
+            update_field(&mut tamagotchi.entertained, FILL_PER_ENTERTAINMENT);
+            msg::reply(TmgEvent::Entertained, 0)
+                .expect("Error sending tamagotchi variant 'Entertained'");  
+        },
+        TmgAction::Sleep => {
+            update_field(&mut tamagotchi.rested, FILL_PER_SLEEP);
+            msg::reply(TmgEvent::Slept, 0)
+                .expect("Error sending tamagotchi variant 'Slept'");  
+        },
+        TmgAction::Transfer(actor_id) => {
+            let source_id = msg::source();
+            let mut owner_transfered = false;
+            if tamagotchi.owner == source_id {
+                tamagotchi.owner = actor_id;
+                owner_transfered = true;
+            }
+            if let Some(approved_account) = tamagotchi.approved_account {
+                if approved_account == source_id {
```

Comment:
You can write it shorter.
```suggestion
            if tamagotchi.approved_account == Some(source_id) {
```